streaming/compression.c

```diff
@@ -0,0 +1,336 @@
+#include "rrdpush.h"
+#include "lz4.h"
+
+#ifdef ENABLE_COMPRESSION
+
+#define LZ4_MAX_MSG_SIZE 0x4000
+#define LZ4_STREAM_BUFFER_SIZE (0x10000 + LZ4_MAX_MSG_SIZE)
+
+#define SIGNATURE ((uint32_t)('z' | 0x80) | (0x80 << 8) | (0x80 << 16) | ('\n' << 24))
+#define SIGNATURE_MASK ((uint32_t)0xff | (0x80 << 8) | (0x80 << 16) | (0xff << 24))
+#define SIGNATURE_SIZE 4
+
+
+/*
+ * LZ4 streaming API compressor specific data
+ */
+struct compressor_data {
+    LZ4_stream_t *stream;
+    char *stream_buffer;
+    size_t stream_buffer_pos;
+};
+
+
+/*
+ * Reset compressor state for a new stream
+ */
+static void lz4_compressor_reset(struct compressor_state *state)
+{
+    if (state->data) {
+        if (state->data->stream)
+            LZ4_resetStream_fast(state->data->stream);
+        state->data->stream_buffer_pos = 0;
+    }
+}
+
+/*
+ * Destroy compressor state and all related data
+ */
+static void lz4_compressor_destroy(struct compressor_state **state)
+{
+    if (state && *state) {
+        struct compressor_state *s = *state;
+        if (s->data) {
+            if (s->data->stream)
+                LZ4_freeStream(s->data->stream);
+            freez(s->data->stream_buffer);
+        }
+        freez(s->buffer);
+        freez(s);
+        *state = NULL;
+    }
+}
+
+/*
+ * Compress the given block of data
+ * Comprecced data will remain in the internal buffer until the next invokation
+ * Return the size of compressed data block as result and the pointer to internal buffer  using the last argument
+ */
+static size_t lz4_compressor_compress(struct compressor_state *state, const char *data, size_t size, char **out)
+{
+    if (!state || !size || !out)
+        return 0;
+    if (size > LZ4_MAX_MSG_SIZE) {
+        error("Message size above limit: %lu", size);
+        return 0;
+    }
```

Comment:
Whenever an error occurs we should differentiate between normal operation and let the call sites handle the error. Maybe changing this to return an `ssize_t` is appropriate here.

Reply:
Hello vkalintiris,
0 is the indication of an error for a calling function.
Unfortunately the calling function sender_commit() has void type and there is no possibility to propogate the error further.

Reply:
> Unfortunately the calling function sender_commit() has void type and there is no possibility to propogate the error further


It's fine to propagate error handling only to `sender_commit`. The reason I'm asking for this is because the compression/decompression stuff is super useful and I can see it "graduating" to it's own folder within `libnetdata` so that it can be used by other components as well in the future.

Reply:
Th require code added in commit 879a451. Please check.